src/replication.c

```diff
@@ -1630,38 +1634,93 @@ int replicationApply(struct raft *r)
 
-    if (index <= r->commit_index) {
-        return;
+    /* In a single voter cluster where this node is part of the configuration,
+     * and is a voter, all stored entries can be committed. */
+    server_index = configurationIndexOf(&r->configuration, r->id);
+    if (configurationVoterCount(&r->configuration) == 1
+        && server_index < r->configuration.n
+        && r->configuration.servers[server_index].role == RAFT_VOTER) {
+        return r->last_stored;
+    }
+
+    /* In all other cases the largest possible new commit index is the largest
+     * match index of another voter in the cluster.
+     * a) In the 1 voter case where this node is no longer part of the
+     *    configuration.
+     * b) In the 1 voter case where this node is no longer a voter.
+     * c) In the multi-voter case when, by definition, at least 1 other voter
+     *    is needed to reach quorum.
+     */
+    for (i = 0, index = 0, max = 0; i < r->configuration.n; i++) {
+        struct raft_server *server = &r->configuration.servers[i];
+        if (server->role != RAFT_VOTER) {
+            continue;
+        }
+        if (server->id == r->id) {
+            continue;
+        }
+        index = r->leader_state.progress[i].match_index;
+        if (index > max) {
+            max = index;
+        }
     }
 
-    /* TODO: fuzzy-test --seed 0x8db5fccc replication/entries/partitioned
-     * fails the assertion below. */
-    if (logTermOf(r->log, index) == 0) {
-        return;
-    }
-    // assert(logTermOf(r->log, index) > 0);
-    assert(logTermOf(r->log, index) <= r->current_term);
+    return max;
+}
 
+static bool matchIndicesAtLeast(struct raft *r, const raft_index index)
+{
+    unsigned votes = 0;
+    unsigned i;
+    tracef("index:%llu count votes", index);
     for (i = 0; i < r->configuration.n; i++) {
         struct raft_server *server = &r->configuration.servers[i];
         if (server->role != RAFT_VOTER) {
+            tracef("server->id:%llu non-voter", server->id);
             continue;
         }
+        tracef("server->id:%llu match_index:%llu", server->id, r->leader_state.progress[i].match_index);
         if (r->leader_state.progress[i].match_index >= index) {
             votes++;
         }
     }
 
-    if (votes > configurationVoterCount(&r->configuration) / 2) {
-        r->commit_index = index;
-        tracef("new commit index %llu", r->commit_index);
+    tracef("index:%llu votes:%u", index, votes);
+    return votes > (configurationVoterCount(&r->configuration) / 2);
+}
+
+void replicationQuorum(struct raft *r)
+{
+    raft_index index;
+    raft_index start_index;
+    raft_term term;
+    assert(r->state == RAFT_LEADER);
+
+    /* Raft Paper fig. 2 - Rules for Servers
+     * If there exists an N such that N > commitIndex, a majority
+     * of matchIndex[i] >= N, and log[N].term == currentTerm:
+     * set commitIndex = N (5.3, 5.4).
+     * */
+    start_index = replicationCalculateCommitIndex(r);
```

Comment:
Yeah, I think it's worth it to optimize this, let me think about it for a bit.

Reply:
I actually think we can keep the current (what's on master) implementation of `replicationQuorum`, while adding the check to not commit entries if the term is older than the current term, on the condition that we call it here https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L754 with `last_index` instead of `r->last_stored`. `last_index` is equal to the `match_index` of the server from which we have just received an AppendEntriesRPC. (due to https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L722)

This works because a new `commit_index` is always equal to some server's `match_index` and all `match_index` updates trickle down to the `replicationQuorum` call. Unless this call https://github.com/canonical/raft/blob/faadebbe8fae5f7faf1f015b90fae094b3fec7e8/src/replication.c#L746 fails, but then we convert to `RAFT_UNAVAILABLE` anyway.

It's a lot simpler solution than the current one proposed in this PR (and still passes the new tests). Let me know what you think.

Reply:
> That reasoning seems sound to me, and I like the simplification. I guess that here
> 
> https://github.com/canonical/raft/blob/0ddb3296fd4ec7d2e8d62b40aa12a0cb3a49e37e/src/replication.c#L520-L521
> 
> we'd still use r->last_stored, right?

yes